Storage/SQL/DeferredDBOperation.swift

```diff
@@ -0,0 +1,111 @@
+
+import Foundation
+import Shared
+import XCGLogger
+
+private let log = XCGLogger.defaultInstance()
+private let DeferredQueue = dispatch_queue_create("BrowserDBQueue", DISPATCH_QUEUE_SERIAL)
+
+/**
+    This class is written to mimick an NSOperation, but also provide Deferred capabilities as well.
+    
+    Usage:
+    let deferred = DeferredDBOperation({ (db, err) -> Int
+      // ... Do something long running
+      return 1
+    }, withDb: myDb, onQueue: myQueue).start(onQueue: myQueue)
+    deferred.upon { res in
+      // if cancelled res.isFailure = true
+    })
+
+    // ... Some time later
+    deferred.cancel()
+*/
+class DeferredDBOperation<T>: Deferred<Result<T>>, Cancellable {
+    /// Cancelled is wrapping a ReadWrite lock to make access to it thread-safe.
+    private var cancelledLock = LockProtected<Bool>(item: false)
+    var cancelled: Bool {
+        get {
+            return self.cancelledLock.withReadLock({ cancelled -> Bool in
+                return cancelled
+            })
+        }
+        set {
+            cancelledLock.withWriteLock { cancelled -> T? in
+                cancelled = newValue
+                return nil
+            }
+        }
+    }
+
+    /// Executing is wrapping a ReadWrite lock to make access to it thread-safe.
+    var connectionLock = LockProtected<SQLiteDBConnection?>(item: nil)
+    var connection: SQLiteDBConnection? {
+        get {
+            // We want to avoid leaking this connection. If you want access to it,
+            // you should use a read/write lock directly.
+            return nil
+        }
+        set {
+            connectionLock.withWriteLock { connection -> T? in
+                connection = newValue
+                return nil
+            }
+        }
+    }
+
+    private var db: SwiftData
+    private var block: (connection: SQLiteDBConnection, inout err: NSError?) -> T
+
+    init(db: SwiftData, block: (connection: SQLiteDBConnection, inout err: NSError?) -> T) {
+        self.block = block
+        self.db = db
+        super.init()
+    }
+
+    func start(onQueue queue: dispatch_queue_t = DeferredQueue) -> DeferredDBOperation<T> {
+        dispatch_async(queue, self.main)
+        return self
+    }
+
+    private func main() {
+        if self.cancelled {
+            let err = NSError(domain: "mozilla", code: 9, userInfo: [NSLocalizedDescriptionKey: "Operation was cancelled before starting"])
+            fill(Result(failure: DatabaseError(err: err)))
+        }
+
+        let start = NSDate.now()
+        var result: T? = nil
+        var err = db.withConnection(SwiftData.Flags.ReadWriteCreate) { (db) -> NSError? in
+            self.connection = db
+            if self.cancelled {
+                return NSError(domain: "mozilla", code: 9, userInfo: [NSLocalizedDescriptionKey: "Operation was cancelled before starting"])
+            }
+
+            var err: NSError? = nil
+            result = self.block(connection: db, err: &err)
```

Comment:
Heh. Swift! I renamed one to error for you so that its a little clearer who you've got in each context.